### PR TITLE
Allow granular updates to user and channel metadata with Optionals

### PR DESCRIPTION
**Pull request by wkal-pubnub**
This allows 3 operations: set, set to null (clear) and don't change. Previously only set and don't change were possible.

### CODE REVIEW

**Review by marcin-cebo — APPROVED** [review]

File: pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/endpoints/objects_api/uuid/SetUserMetadata.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.pubnub.api.endpoints.objects_api.uuid
+
+import com.pubnub.api.endpoints.Endpoint
+import com.pubnub.api.models.consumer.objects_api.uuid.PNSetUUIDMetadataResult
+import com.pubnub.api.utils.Optional
+
+interface SetUserMetadata : Endpoint<PNSetUUIDMetadataResult?> {
+    fun custom(custom: Optional<Map<String, Any?>?>): SetUserMetadata
+
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/endpoints/objects_api/uuid/SetUserMetadata.kt:9]: no it's not, it's whatever there was before in the file I copied it from
done

**Review by marcin-cebo — APPROVED** [review]

File: pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/PubNub.kt [hunk 2]
```diff
@@ -442,10 +443,23 @@ interface PubNub :
     fun removeUUIDMetadata(): RemoveUUIDMetadata
 
     /**
-     * Set metadata for a Channel in the database, optionally including the custom data object for each.
+     * Set metadata for a Channel in the database, optionally including the custom data object.
      */
+    @Deprecated(
+        message = "It's not possible to set metadata to `null` using this function. Use `updateChannelMetadata` " +
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/PubNub.kt:449]: the other method covers a superset of cases, therefore this method is not needed or preferred in any circumstances. we should remove it in the future

WARNING is the default, no need to specify it

File: pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/PubNubImpl.java [hunk 3]
```diff
@@ -374,10 +376,17 @@ public RemoveUUIDMetadata removeUUIDMetadata() {
...
 
     @Override
+    @NotNull
     public SetChannelMetadata.Builder setChannelMetadata() {
         return new SetChannelMetadataImpl.Builder(getPubNubCore());
     }
 
+    @Override
+    @NotNull
+    public UpdateChannelMetadata.Builder updateChannelMetadata() {
```
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/PubNubImpl.java:386]: done

**Review by marcin-cebo — APPROVED** [review]

**Comment by wkal-pubnub** [comment]
Putting this PR on hold until major release. 
TBD: instead of deprecating old methods and having ones with new names, we will replace the old methods parameters with new ones (based on Optional), which will be a breaking change.

**Pull request closed without merging**